rfcs/20200618-headers/README.md

```diff
@@ -0,0 +1,123 @@
+# Changes in the oneDNN header install locations and namespace
+
+## Introduction
+
+New headers structure and namespaces are needed for cross-component and oneAPI
+branding alignment. The request from the oneAPI program is to put oneDNN
+headers under the `oneapi` and also to provide a corresponding namespace. The
+changes will have to be reflected in the
+[spec](https://github.com/oneapi-src/oneAPI-spec/tree/master/source/elements/oneDNN).
+
+The main concern is how to keep backward compatibility with the existing code.
+The secondary concern is maintenance burden.
+
+## Proposal
+
+The following changes are to be implemented in the
+[`dev-v2`](https://github.com/oneapi-src/onednn/tree/dev-v2) branch but will
+eventually be part of master.
+
+### Namespace changes
+
+oneDNN should start providing two namespaces (C++ only): `dnnl::` and
+`oneapi::dnnl::`. Either can be an alias. To minimize the changes, the
+proposal is to alias the long one:
+
+```c++
+namespace dnnl { /* the usual stuff */ }
+
+namespace oneapi { namespace dnnl = ::dnnl; }
+```
+
+The new namespace must be reflected in the
+[spec](https://github.com/oneapi-src/oneAPI-spec/tree/master/source/elements/oneDNN).
+There is no recommendation on whether longer or shorted namespace should be
+used in the spec yet; this is outside of the scope of this RFC.
+
+### Installation paths changes
+
+oneDNN is free to decide on the header structure under `include/oneapi`.
+
+#### Option 1 -- recommended
+
+```
+include
+├── oneapi[/dnnl]
+│   ├── dnnl_types.h
+│   ├── dnnl_config.h
+│   ├── dnnl.hpp
+│   └── dnnl.h
+├── dnnl_types.h        (#include "oneapi[/dnnl]/dnnl_types.h")
+├── dnnl_config.h       (#include "oneapi[/dnnl]/dnnl_config.h")
+├── dnnl.hpp            (#include "oneapi[/dnnl]/dnnl.hpp")
+└── dnnl.h              (#include "oneapi[/dnnl]/dnnl.h")
+```
+
+The `/dnnl/` portion of the path may be present or not. The recommendation is
+*to include it* so that it is possible to extend dnnl library later in a
+straightforward manner.
+
+Additionally, the compatibility headers can complain like this:
+
+```c++
+#ifdef _MSC_VER
+#pragma message "Warning: this header is deprecated. Please include oneapi[/dnnl]/dnnl.hpp instead"
+#else
+#warning "This header is deprecated. Please include oneapi[/dnnl]/dnnl.hpp instead"
+#endif
+
+#include "oneapi[/dnnl]/dnnl.hpp"
+```
+
+The warnings can be suppressed via `-Wno-cpp` (clang, gcc),
+`-diag-disable:1224` (icc). The MSVC message cannot be disabled, but it does
+not show as a warning, just a message. (However, the internal oneDNN CI will
+still treat them as warning because the word 'warning' appears in the
+message.) However, it is *not recommended* to not cause additional confusion.
+The deprecation will be noted in the documentation.
+
+This option *is recommended*. The canonical include path to `dnnl.hpp` will
+become `#include "oneapi/dnnl/dnnl.hpp"`.
```

Comment:
The recommended pattern is `#include "oneapi/dnnl.hpp"` but `#include "oneapi/dnnl/feature.hpp"`, assuming `dnnl.hpp` is the all-in-one header but you might also have self-contained features that allow independent usage (e.g. one candidate could be `#include "oneapi/dnnl/version.hpp"` for just version checks. Note that with the directory you do not have to use `dnnl_` as a file name prefix.

Reply:
As dneDNN does not really have feature headers, do you think it would be acceptable to have `#include "oneapi/dnnl/dnnl.hpp"`? I can certainly think about splitting the `dnnl.hpp` but that is out of scope at the moment.